### PR TITLE
difference and complement

### PR DESCRIPTION
**Pull request by MareoRaft**
**Summary**

PR as per issue #1 

**Description**

change `complement` to `difference` (relative complement).  Modify `complement` to enforce that set1 must be the entire universe (absolute complement).

**Notes**

This is not yet ready for merge.

  1. I wasn't able to run the tests.  I tried running them with `jasmine index.tests.js`, but I got the error

    /Users/Matthew/programming/js-set-extension/package/lib/index.tests.js:28
    import './index'
    ^^^^^^
    
    SyntaxError: Unexpected token import

  2. I filled out the TODOs for testing the complement, but I didn't fill out the ones for difference yet.  Once I get the tests working for complement, I'll write the ones for difference too.

  3. Is there a @seealso directive supported by your docs?  It might be nice to allow users viewing `difference` to be pointed to `complement`, and vice versa, since they are so similar.

### CODE REVIEW

File: package/lib/index.js [hunk 1]
```diff
@@ -454,7 +454,7 @@ global.Set.prototype = _originalSet.prototype
 // //////////////////////////////////////////////////////////////////////////////// //
 
 /**
- * Creates a new Set from arbitrary arguments wihtout the need of "new" and the array notation.
+ * Creates a new Set from arbitrary arguments without the need of "new" and the array notation.
```
Comment (MareoRaft) [inline package/lib/index.js:457]: minor grammatical improvements

File: package/lib/index.js [hunk 2]
```diff
@@ -573,6 +573,28 @@ function complement (set1, set2) {
...
+ * @throws Throws an error if any element in B does not occur in U.
+ * @param set1 - U the set to be subtracted from
+ * @param set2 - B the set whose elements will be subtracted from A
+ * @returns {ExtendedSet|*} A new Set with all elements of U minus the elements of B
+ */
+function complement (set1, set2) {
+  checkSet(set1)
+  checkSet(set2)
+  if (!set1.isSupersetOf(set2)) {
+    throw new Error(`[set2] has an element which is not in the universe [se1].`)
```
Comment (MareoRaft) [inline package/lib/index.js]: Enforce that set1 must be the entire universe

File: package/lib/index.tests.js [hunk 3]
```diff
@@ -671,32 +671,32 @@ describe('Operations (static)', function () {
     })
   })
 
-  describe(Set.complement.name, function () {
-    // used with https://en.wikipedia.org/wiki/Complement_(set_theory)
+  describe(Set.difference.name, function () {
+    // used with https://en.wikipedia.org/wiki/Complement_(set_theory)#Relative_complement
```
Comment (MareoRaft) [inline package/lib/index.tests.js:680]: More specific wikipedia link

File: package/lib/index.tests.js [hunk 4]
```diff
@@ -709,6 +709,96 @@ describe('Operations (static)', function () {
...
+      }, /Expected \[set\] to be instanceof \[Set\]/)
+
+      assert.throws(function () {
+        Set.difference(set(1), null)
+      }, /Expected \[set\] to be instanceof \[Set\]/)
+    })
+  })
+
+  describe(Set.complement.name, function () {
+    // used with https://en.wikipedia.org/wiki/Complement_(set_theory)#Absolute_complement
```
Comment (MareoRaft) [inline package/lib/index.tests.js:783]: complement stuff here

File: package/lib/index.tests.js [hunk 5]
```diff
@@ -709,6 +709,96 @@ describe('Operations (static)', function () {
...
+
+      // {1, 2, 3, 4} \ {1, 3} = {2, 4}.
+      areEqual(Set.complement(c, b), new Set([2, 4]))
+
+      // If B has an element not in A (impossible since A must be the entire universe), throw an error.
+      assert.throws(function() {
+        Set.complement(a, b)
+      }, /\[set2\] has an element which is not in the universe \[se1\]\./)
+
+      // A \ ∅ = A.
```
Comment (MareoRaft) [inline package/lib/index.tests.js:804]: filled out 'TODO' tests, but removed a few which are illegal for the absolute complement

**Comment by jankapunkt** [comment]
You can run the tests like the following:

```bash
$ cd js-set-extension/package
$ npm install
```

To run tests in watch mode use

```bash
$ npm run test-watch
```

or for a single run use

```bash
$ npm run test
```

The tests are written in mocha but it should not be that hard to get into it as it is written very intuitive.

## Linting

Please note, that the tests are very strict about code style and you can check for code style related errors using

```bash
npm run lint
```

You should fix these lint errors, since the CI server will reject to run any tests when the linter as thrown an error.

You can also run lint and tests all in one process using

```bash
npm run lint-test
```

The scripts to run these commands are also in the package.json file:

https://github.com/jankapunkt/js-set-extension/blob/master/package/package.json

# How to update the code to this PR

If you make updates to your forked repo on the branch you have opened this PR (it is `MareoRaft:master` as you can see in the title) it will automatically be added to this PR and the CI server will also run the tests again. So no worries in changing and updating your code here.

**Comment by jankapunkt** [comment]
@MareoRaft btw just let me know, if you need more help with the testing and linting.

**Comment by MareoRaft** [comment]
This should pretty much be good to go now.  Thanks for your help!

File: .gitignore [hunk 6]
```diff
@@ -62,3 +62,5 @@ typings/
 
 # IDE specific
 .idea
+*.sublime-project
+*.sublime-workspace
```
Comment (MareoRaft) [inline .gitignore:66]: added this so repo would ignore files for my personal IDE / text editor.  Can remove this if you don't want it.

File: package/lib/index.tests.js [hunk 7]
```diff
@@ -543,13 +544,17 @@ describe('Operations (static)', function () {
       // A ∪ A = A.
       assert.isTrue(Set.union(a, a).equal(a))
 
-      // A ∪ U = U. // TODO even possible to test here? -> https://en.wikipedia.org/wiki/Russell%27s_paradox
+      // A ∪ U = U.
+      areEqual(u, Set.union(a, u))
```
Comment (MareoRaft) [inline package/lib/index.tests.js:548]: Russell's paradox says that there is no such thing as "the set of all sets".  I believe you are equating this to "the universe".

However, these are two separate concepts.  When working within a "universe", it is in fact *you* who gets to choose which universe you want to work within.  For example, I have chosen for my universe to be {1, 2, 3, 4, 5, 6, 7}.  All this really means is that all other sets (such as a, b, and c) cannot possibly contain elements outside of {1, 2, 3, 4, 5, 6, 7}.

File: package/lib/index.tests.js [hunk 8]
```diff
@@ -671,42 +676,59 @@ describe('Operations (static)', function () {
...
 
       // ∅ \ A = ∅.
-      assert.equal(Set.complement(e, a).size, 0)
+      assert.equal(Set.difference(e, a).size, 0)
 
       // A \ ∅ = A.
-      areEqual(Set.complement(a, e), a)
+      areEqual(Set.difference(a, e), a)
 
-      // TODO
```
Comment (MareoRaft) [inline package/lib/index.tests.js:701]: finished all TODOs for difference below

File: .gitignore (hunk 6 above)
Comment (jankapunkt) [reply .gitignore:66]: It's good to keep it

File: package/lib/index.tests.js (hunk 8 above)
Comment (jankapunkt) [reply package/lib/index.tests.js:701]: Thanks a lot for finishing them!

File: package/lib/index.tests.js (hunk 5 above)
Comment (jankapunkt) [reply package/lib/index.tests.js:804]: Yes, you mentioned this in the PR. Thanks a lot!

**Review by jankapunkt — APPROVED** [review]

**Pull request merged**